vllm/distributed/device_communicators/quick_all_reduce.py

```
@@ -0,0 +1,122 @@
+        rank = dist.get_rank(group=self.group)
+        world_size = dist.get_world_size(group=self.group)
+        if world_size == 1:
+            # No need to initialize QuickReduce for single GPU case.
+            return
+
+        if world_size not in QuickAllReduce._SUPPORTED_WORLD_SIZES:
+            logger.warning(
+                "QuickReduce allreduce is disabled due to an unsupported world"
+                " size: %d. Supported world sizes: %s."
+                " To disable this warning set VLLM_ROCM_CA_BACKEND"
+                " to None", world_size,
+                str(QuickAllReduce._SUPPORTED_WORLD_SIZES))
+            return
+
+        assert current_platform.is_rocm(), (
+            "QuickReduce is only supported on ROCm platform.")
+        if isinstance(device, int):
+            device = torch.device(f"cuda:{device}")
+        elif isinstance(device, str):
+            device = torch.device(device)
+        # now `device` is a `torch.device` object
+        assert isinstance(device, torch.device)
+        self.device = device
+        torch.cuda.set_device(self.device)
+
+        self._ptr = ops.init_custom_qr(rank, world_size)
+        self.create_shared_buffer()
+        self.disabled = False
+
+    def create_shared_buffer(self):
+        """
+        Creates a shared buffer for quickreduce. 
+        Has to be called after qr_init_device_collectives
+        """
+        handle = ops.qr_get_handle(self._ptr)
+        world_size = dist.get_world_size(group=self.group)
+        handles = [None] * world_size
+        dist.all_gather_object(handles, handle, group=self.group)
+        ops.qr_open_handles(self._ptr, handles)
+
+    def all_reduce(self, inp: torch.Tensor, *, out: torch.Tensor = None):
+        """Performs an out-of-place all reduce.       
+        """
+        inp_size = inp.numel() * inp.element_size()
+        if inp_size >= self.max_size:
+            return None
```

This should raise an exception if `quick_allreduce` is invalid instead of silently returning None